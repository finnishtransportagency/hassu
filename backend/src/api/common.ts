import {
  LataaProjektiQueryVariables,
  ListaaVelhoProjektitQueryVariables,
  TallennaProjektiInput,
  TilaSiirtymaInput,
} from "../../../common/graphql/apiModel";
import { AppSyncResolverEvent } from "aws-lambda/trigger/appsync-resolver";
import { getCorrelationId, setupLambdaMonitoring, setupLambdaMonitoringMetaData, wrapXRayAsync } from "../aws/monitoring";
import { log, setLogContextOid } from "../logger";
import { identifyUser } from "../user";
import { ClientError } from "../error/ClientError";
import { SystemError } from "../error/SystemError";
import { IllegalAineistoStateError } from "../error/IllegalAineistoStateError";
import { NotFoundError } from "../error/NotFoundError";
import { IllegalAccessError } from "../error/IllegalAccessError";

export type AppSyncEventArguments =
  | unknown
  | LataaProjektiQueryVariables
  | ListaaVelhoProjektitQueryVariables
  | TallennaProjektiInput
  | TilaSiirtymaInput;

export type LambdaResult = {
  data: unknown;
  correlationId: string;
};

export async function commonHandleEvent(
  event: AppSyncResolverEvent<AppSyncEventArguments>,
  executeOperation: (event: AppSyncResolverEvent<AppSyncEventArguments>) => Promise<unknown>
): Promise<LambdaResult> {
  setupLambdaMonitoring();
  if (log.isLevelEnabled("debug")) {
    log.debug({ event });
  }

  return wrapXRayAsync("handler", async (subsegment) => {
    setupLambdaMonitoringMetaData(subsegment);

    try {
      await identifyUser(event);
      const data = await executeOperation(event);
      const corId = getCorrelationId();
      const lambdaResult: LambdaResult = { data, correlationId: corId || "" };
      return lambdaResult;
    } catch (e: unknown) {
      if (e instanceof Error) {
        // Only data that is sent out in case of error is the error message. We wish to log correlationId with the
        // error, so the only way to do it is to encode the data into error message field. The error field is decoded
        // in deployment/lib/template/response.vtl

        let errorType = "Error";
        let errorSubType = "(no subtype)";
        if (
          e instanceof NotFoundError ||
          e instanceof IllegalAineistoStateError ||
          e instanceof IllegalAccessError ||
          e instanceof ClientError
        ) {
          errorType = "ClientError";
          errorSubType = e.className;
          log.error(e.message);
        } else if (e instanceof SystemError) {
          errorType = "SystemError";
          errorSubType = e.className;
          log.error(e);
        } else {
          log.error(e);
        }

        e.message = JSON.stringify({
          message: e.message,
          correlationId: getCorrelationId(),
          errorType,
          errorSubType,
        });
      }
      throw e;
    } finally {
      setLogContextOid(undefined);
    }
  });
}
