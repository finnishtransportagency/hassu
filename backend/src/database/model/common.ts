import {
  AineistoTila,
  ELY,
  EVK,
  IlmoitettavaViranomainen,
  Kieli,
  ProjektiTyyppi,
  SuunnittelustaVastaavaViranomainen,
  LadattuTiedostoTila,
} from "hassu-common/graphql/apiModel";

export type LocalizedMap<T> = { [key in Kieli]?: T } | null;
export type RequiredLocalizedMap<T> = {
  [Kieli.SUOMI]: T;
  [Kieli.RUOTSI]?: T;
};

export enum SaameKieli {
  POHJOISSAAME = "POHJOISSAAME",
}

export enum SuomiRuotsiKieli {
  SUOMI = "SUOMI",
  RUOTSI = "RUOTSI",
}

export type SaameLocalizedMap<T> = { [key in SaameKieli]?: T } | null;

export type Kielitiedot = {
  ensisijainenKieli: Kieli;
  toissijainenKieli?: Kieli | null;
  projektinNimiVieraskielella?: string;
};

export type Yhteystieto = {
  etunimi: string;
  sukunimi: string;
  organisaatio?: string;
  kunta?: number | null;
  elyOrganisaatio?: ELY;
  evkOrganisaatio?: EVK;
  puhelinnumero: string;
  sahkoposti: string;
  titteli?: string;
};

export type Aineisto = {
  // Dokumentin oid Velhossa
  dokumenttiOid: string;
  // Kategorian ID, joka viittaa kategoriapuun kategoriaan
  kategoriaId?: string;
  // Suhteellinen polku tiedostoon yllapidon S3-bucketissa projektin alla
  tiedosto?: string;
  // Tiedostonimi naytettavaksi
  nimi: string;
  // Tunniste, joka on luotu FE:ssä, kun tiedosto on valittu. Ei muutu.
  uuid: string;
  // Jos tosi, tallennuksen validoinnissa ja tallennuksessa käytetään tunnisteena uuid:n sijaa dokumenttiOid:ta
  // HUOM! tätä tietoa ei ole oikeasti tarkoitus tallentaa kantaan
  uuidGeneratedBySchemaMigration?: boolean;
  // Aikaleima, milloin tiedosto on tuotu jarjestelmaan yyyy-MM-ddTHH:mm
  tuotu?: string;
  // Numero jarjestamista varten
  jarjestys?: number | null;
  tila: AineistoTila;
  kategoriaMuuttunut?: boolean;
};

export type AineistoNew = ILadattuTiedosto & {
  // Dokumentin oid Velhossa
  dokumenttiOid: string;
  // Kategorian ID, joka viittaa kategoriapuun kategoriaan
  kategoriaId?: string;
};

export type StandardiYhteystiedot = {
  yhteysTiedot?: Yhteystieto[];
  yhteysHenkilot?: string[];
};

export type IlmoituksenVastaanottajat = {
  kunnat?: Array<KuntaVastaanottaja> | null;
  viranomaiset?: Array<ViranomaisVastaanottaja> | null;
  maakunnat?: Array<MaakuntaVastaanottaja> | null;
};

export type SahkopostiVastaanottaja = {
  sahkoposti: string;
  messageId?: string;
  lahetetty?: string | null;
  lahetysvirhe?: boolean | null;
};

export type KuntaVastaanottaja = {
  id: number;
} & SahkopostiVastaanottaja;

export type MaakuntaVastaanottaja = {
  id: number;
} & SahkopostiVastaanottaja;

export type ViranomaisVastaanottaja = {
  nimi: IlmoitettavaViranomainen;
} & SahkopostiVastaanottaja;

export type UudelleenKuulutus = {
  tila: UudelleenkuulutusTila;
  selosteKuulutukselle?: RequiredLocalizedMap<string>;
  selosteLahetekirjeeseen?: RequiredLocalizedMap<string>;
  alkuperainenHyvaksymisPaiva?: string;
  tiedotaKiinteistonomistajia?: boolean;
  alkuperainenKuulutusPaiva?: string;
};

export type AineistoMuokkaus = {
  alkuperainenHyvaksymisPaiva?: string;
};

export enum UudelleenkuulutusTila {
  PERUUTETTU = "PERUUTETTU",
  JULKAISTU_PERUUTETTU = "JULKAISTU_PERUUTETTU",
}

export type LinkitettyVelhoProjekti = {
  oid: string;
  nimi: string;
  tyyppi?: ProjektiTyyppi | null;
};

export type Velho = {
  nimi: string;
  tyyppi?: ProjektiTyyppi | null;
  kuvaus?: string | null;
  vaylamuoto?: string[] | null;
  asiatunnusVayla?: string | null;
  asiatunnusELY?: string | null;
  asiatunnusEVK?: string | null;
  suunnittelustaVastaavaViranomainen?: SuunnittelustaVastaavaViranomainen | null;
  toteuttavaOrganisaatio?: string | null;
  vastuuhenkilonNimi?: string | null;
  vastuuhenkilonEmail?: string | null;
  varahenkilonNimi?: string | null;
  varahenkilonEmail?: string | null;
  maakunnat?: number[] | null;
  kunnat?: number[] | null;
  linkki?: string | null;
  linkitetytProjektit?: LinkitettyVelhoProjekti[] | null;
  geoJSON?: string | null;
};
export interface ILadattuTiedosto {
  // Kayttajalle esitettava tiedostonimi
  nimi: string;
  uuid: string;
  lisatty: string;
}

export type LadattuTiedostoNew = ILadattuTiedosto;

export type LadattuTiedosto = {
  // Suhteellinen polku tiedostoon yllapidon S3-bucketissa projektin alla
  tiedosto: string;
  // Kayttajalle esitettava tiedostonimi
  nimi?: string | null;
  uuid: string;
  tuotu?: string | null;
  //Numero jarjestamista varten
  jarjestys?: number;
  tila: LadattuTiedostoTila;
};

export type KunnallinenLadattuTiedosto = {
  kunta: number;
} & ILadattuTiedosto;

export type Laskutustiedot = {
  ovtTunnus?: string | null;
  verkkolaskuoperaattorinTunnus?: string | null;
  viitetieto?: string | null;
};
