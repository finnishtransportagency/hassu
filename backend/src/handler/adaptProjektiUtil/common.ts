import * as API from "../../../../common/graphql/apiModel";

export function adaptIlmoituksenVastaanottajat(
  vastaanottajat: API.IlmoituksenVastaanottajat | null | undefined
): API.IlmoituksenVastaanottajat {
  if (!vastaanottajat) {
    return vastaanottajat as null | undefined;
  }
  const kunnat: API.KuntaVastaanottaja[] =
    vastaanottajat?.kunnat?.map((kunta) => ({ __typename: "KuntaVastaanottaja", ...kunta })) || null;
  const viranomaiset: API.ViranomaisVastaanottaja[] =
    vastaanottajat?.viranomaiset?.map((viranomainen) => ({
      __typename: "ViranomaisVastaanottaja",
      ...viranomainen,
    })) || null;
  return { __typename: "IlmoituksenVastaanottajat", kunnat, viranomaiset };
}
