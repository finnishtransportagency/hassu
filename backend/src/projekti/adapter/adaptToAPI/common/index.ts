export * from "./lisaaTypename";
export * from "./adaptLadattuTiedostoToAPI";
export * from "./adaptKuulutusSaamePDFtToAPI";
export * from "./adaptLokalisoituTekstiToAPI";
export * from "./adaptPakotettuLokalisoituTekstiToAPI";
export * from "./adaptLogotToAPI";
export * from "./adaptUudelleenKuulutusToAPI";
export * from "./adaptAineistoMuokkausToAPI";
export * from "./adaptAineistotToAPI";
export * from "./adaptIlmoituksenVastaanottajatToAPI";
export * from "./adaptLokalisoituLinkkiToAPI";
export * from "./adaptStandardiYhteystiedotToAPIByAddingProjektiPaallikko";
export * from "./adaptVelhoToAPI";
