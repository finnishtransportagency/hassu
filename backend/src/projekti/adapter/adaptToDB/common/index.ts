export * from "./adaptTiedostotToSave";
export * from "./removeTypeName";
export * from "./adaptLogoFilesToSave";
export * from "./adaptIlmoituksenVastaanottajatToSave";
export * from "./adaptYhteystiedotToSave";
export * from "./adaptAineistotToSave";
export * from "./adaptHankkeenKuvausToSave";
export * from "./adaptLokalisoituTekstiToSave";
export * from "./adaptLokalisoituLinkkiToSave";
export * from "./getId";
export * from "./forEverySaameDoAsync";
export * from "./adaptLadattuTiedostoToSave";
