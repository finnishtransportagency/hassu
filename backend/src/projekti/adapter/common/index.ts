export * from "./adaptAineistot";
export * from "./adaptIlmoituksenVastaanottajat";
export * from "./util";
export * from "./lisaaTypename";
export * from "./adaptHankkeenKuvaus";
