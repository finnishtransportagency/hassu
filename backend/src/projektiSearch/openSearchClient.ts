import { HttpRequest } from "@aws-sdk/protocol-http";
import { openSearchConfig } from "./openSearchConfig";
import { sendSignedRequest } from "../aws/awsRequest";
import { log } from "../logger";

const domain = openSearchConfig.searchDomain ?? "search-domain-missing";
const type = "_doc";

export async function sendRequest(request: HttpRequest): Promise<unknown> {
  return (await sendSignedRequest(request, "es")).body;
}

export interface SortOrder {
  order: "asc" | "desc";
}

export enum OpenSearchIndexType {
  YLLAPITO = "yllapito",
  JULKINEN = "julkinen",
}

export interface SearchOpts {
  from?: number;
  size?: number;
  sort?: Record<string, SortOrder>[];
  query?: unknown;
  aggs?: unknown;
}

export default class OpenSearchClient {
  private index: string;

  constructor(index: string) {
    this.index = index;
  }

  async putDocument(key: string, document: unknown): Promise<void> {
    const response = (await this.put("/" + type + "/" + key, JSON.stringify(document))) as { result: string };
    if (response["result"] !== "created" && response["result"] !== "updated") {
      log.warn(response);
    }
  }

  async deleteDocument(key: string): Promise<void> {
    await this.delete("/" + type + "/" + key);
  }

  async delete(path: string): Promise<unknown> {
    const request = new HttpRequest({
      headers: {
        host: domain,
      },
      hostname: domain,
      method: "DELETE",
      path: this.index + path,
    });
    return sendRequest(request);
  }

  async deleteIndex(): Promise<unknown> {
    return this.delete("");
  }

  async createIndex(settings: any, mappings: any): Promise<unknown> {
    const body = JSON.stringify({ settings, mappings });
    return this.put("", body);
  }

  async indexExists(): Promise<boolean> {
    const request = new HttpRequest({
      headers: {
        host: domain,
      },
      query: { pretty: "true" },
      hostname: domain,
      method: "HEAD",
      path: this.index,
    });
    try {
      const response = await sendSignedRequest(request, "es");
      return response.statusCode === 200;
    } catch (e) {
      log.info("No index found", e);
    }
    return false;
  }

  async query(query: SearchOpts): Promise<any> {
    const body = JSON.stringify(query);
    log.info("query " + this.index, { query });
    const request = new HttpRequest({
      body,
      headers: {
        "Content-Type": "application/json",
        host: domain,
      },
      hostname: domain,
      method: "POST",
      path: this.index + "/_search",
    });
    return sendRequest(request);
  }

  async getSettings(): Promise<unknown> {
    return this.get("/_settings");
  }

  async putSettings(settings: string): Promise<unknown> {
    await this.closeIndex();

    try {
      return await this.put("/_settings", settings);
    } finally {
      await this.openIndex();
    }
  }

  async getMapping(): Promise<unknown> {
    const mapping = await this.get("/_mapping");
    return (mapping as any).projekti.mappings;
  }

  async putMapping(mapping: string): Promise<unknown> {
    return this.put("/_mapping", mapping);
  }

  async get(path: string): Promise<any> {
    const request = new HttpRequest({
      headers: {
        host: domain,
      },
      query: { pretty: "true" },
      hostname: domain,
      method: "GET",
      path: this.index + path,
    });
    return sendRequest(request);
  }

  async put(path: string, body: string): Promise<unknown> {
    const request = new HttpRequest({
      headers: {
        "Content-Type": "application/json",
        host: domain,
      },
      body,
      hostname: domain,
      method: "PUT",
      path: this.index + path,
    });
    return sendRequest(request);
  }

  protected async closeIndex(): Promise<void> {
    log.info("closeIndex " + this.index);
    const request = new HttpRequest({
      headers: {
        "Content-Type": "application/json",
        host: domain,
      },
      hostname: domain,
      method: "POST",
      path: this.index + "/_close",
    });
    log.info(await sendRequest(request));
  }

  private async openIndex() {
    log.info("openIndex " + this.index);
    const request = new HttpRequest({
      headers: {
        "Content-Type": "application/json",
        host: domain,
      },
      hostname: domain,
      method: "POST",
      path: this.index + "/_open",
    });
    log.info(sendRequest(request));
  }
}
