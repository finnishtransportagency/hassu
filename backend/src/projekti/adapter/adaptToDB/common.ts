import * as API from "../../../../../common/graphql/apiModel";
import { IllegalArgumentError } from "../../../error/IllegalArgumentError";
import { Aineisto, LocalizedMap } from "../../../database/model";
import { AineistoChangedEvent, ProjektiAdaptationResult, ProjektiEventType } from "../projektiAdapter";
import remove from "lodash/remove";

export function adaptIlmoituksenVastaanottajatToSave(
  vastaanottajat: API.IlmoituksenVastaanottajatInput | null | undefined
): API.IlmoituksenVastaanottajat {
  if (!vastaanottajat) {
    return vastaanottajat as null | undefined;
  }
  const kunnat: API.KuntaVastaanottaja[] = vastaanottajat?.kunnat?.map((kunta) => ({ __typename: "KuntaVastaanottaja", ...kunta })) || null;
  if (!vastaanottajat?.viranomaiset || vastaanottajat.viranomaiset.length === 0) {
    throw new IllegalArgumentError("Viranomaisvastaanottajia pitää olla vähintään yksi.");
  }
  const viranomaiset: API.ViranomaisVastaanottaja[] =
    vastaanottajat?.viranomaiset?.map((viranomainen) => ({
      __typename: "ViranomaisVastaanottaja",
      ...viranomainen,
    })) || null;
  return { __typename: "IlmoituksenVastaanottajat", kunnat, viranomaiset };
}

export function adaptYhteystiedotToSave(yhteystietoInputs: Array<API.YhteystietoInput>): API.YhteystietoInput[] | undefined {
  return yhteystietoInputs?.length > 0 ? yhteystietoInputs.map((yt) => ({ ...yt })) : undefined;
}

export function adaptYhteysHenkilotToSave(yhteystiedot: string[]): string[] {
  return yhteystiedot.filter((yt, index) => yhteystiedot.indexOf(yt) === index);
}

export function adaptStandardiYhteystiedot(kuulutusYhteystiedot: API.StandardiYhteystiedotInput): API.StandardiYhteystiedotInput {
  return {
    yhteysTiedot: adaptYhteystiedotToSave(kuulutusYhteystiedot.yhteysTiedot),
    yhteysHenkilot: adaptYhteysHenkilotToSave(kuulutusYhteystiedot.yhteysHenkilot),
  };
}

export function adaptHankkeenKuvausToSave(hankkeenKuvaus: API.HankkeenKuvauksetInput): LocalizedMap<string> {
  if (!hankkeenKuvaus) {
    return undefined;
  }
  return { ...hankkeenKuvaus };
}

export function adaptAineistotToSave(
  dbAineistot: Aineisto[] | undefined,
  aineistotInput: API.AineistoInput[] | undefined,
  projektiAdaptationResult: Partial<ProjektiAdaptationResult>
): Aineisto[] | undefined {
  const resultAineistot = [];
  let hasPendingChanges = undefined;

  // Examine and update existing documents
  if (dbAineistot) {
    dbAineistot.forEach((dbAineisto) => {
      const updateAineistoInput = pickAineistoFromInputByDocumenttiOid(aineistotInput, dbAineisto.dokumenttiOid);
      if (updateAineistoInput) {
        // Update existing one

        if (dbAineisto.nimi !== updateAineistoInput.nimi) {
          hasPendingChanges = true;
        }
        dbAineisto.nimi = updateAineistoInput.nimi;
        dbAineisto.jarjestys = updateAineistoInput.jarjestys;
        dbAineisto.kategoriaId = updateAineistoInput.kategoriaId;
        resultAineistot.push(dbAineisto);
      }
      if (!updateAineistoInput && aineistotInput) {
        dbAineisto.tila = API.AineistoTila.ODOTTAA_POISTOA;
        resultAineistot.push(dbAineisto);
        hasPendingChanges = true;
      }
    });
  }

  // Add new ones and optionally trigger import later
  if (aineistotInput) {
    for (const aineistoInput of aineistotInput) {
      resultAineistot.push({
        dokumenttiOid: aineistoInput.dokumenttiOid,
        nimi: aineistoInput.nimi,
        kategoriaId: aineistoInput.kategoriaId,
        jarjestys: aineistoInput.jarjestys,
        tila: API.AineistoTila.ODOTTAA_TUONTIA,
      });
      hasPendingChanges = true;
    }
  }

  if (hasPendingChanges) {
    projektiAdaptationResult.pushEvent({ eventType: ProjektiEventType.AINEISTO_CHANGED } as AineistoChangedEvent);
  }
  return resultAineistot;
}

function pickAineistoFromInputByDocumenttiOid(aineistotInput: API.AineistoInput[], dokumenttiOid: string) {
  const matchedElements = remove(aineistotInput, { dokumenttiOid });
  if (matchedElements.length > 0) {
    return matchedElements[0];
  }
  return undefined;
}
