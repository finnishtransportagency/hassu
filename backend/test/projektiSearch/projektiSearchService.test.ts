import { describe, it } from "mocha";
import { openSearchClientJulkinen, openSearchClientYllapito } from "../../src/projektiSearch/openSearchClient";
import { projektiSearchService } from "../../src/projektiSearch/projektiSearchService";
import * as sinon from "sinon";
import { Kieli, ProjektiTyyppi, Status, Viranomainen } from "../../../common/graphql/apiModel";
import { kuntametadata } from "../../../common/kuntametadata";

const sandbox = require("sinon").createSandbox();

const { expect } = require("chai");

const fakeSearchResponse = {
  took: 8,
  timed_out: false,
  _shards: { total: 5, successful: 5, skipped: 0, failed: 0 },
  hits: {
    total: { value: 2, relation: "eq" },
    max_score: 1.0,
    hits: [
      {
        _index: "projekti",
        _type: "_doc",
        _id: "1",
        _score: 1.0,
        _source: {
          muistiinpano: "testi",
          tyyppi: "TIE",
          velho: {
            tilaajaOrganisaatio: "Etelä-Pohjanmaan ELY-keskus",
            nimi: "Nimi1",
            vaylamuoto: ["tie"],
          },
          status: "EI_JULKAISTU",
        },
      },
      {
        _index: "projekti",
        _type: "_doc",
        _id: "2",
        _score: 1.0,
        _source: {
          muistiinpano: "",
          tyyppi: "TIE",
          velho: {
            tilaajaOrganisaatio: "Uudenmaan ELY-keskus",
            nimi: "Nimi2",
            vaylamuoto: ["tie"],
          },
          status: "EI_JULKAISTU",
        },
      },
    ],
  },
};
describe("ProjektiSearchService", () => {
  let openSearchQueryStub: sinon.SinonStub;
  let openSearchSuomiQueryStub: sinon.SinonStub;
  beforeEach(() => {
    openSearchQueryStub = sandbox.stub(openSearchClientYllapito, "query");
    openSearchSuomiQueryStub = sandbox.stub(openSearchClientJulkinen["SUOMI"], "query");
  });
  afterEach(() => {
    sandbox.reset();
    sandbox.restore();
  });

  it("should handle query errors", async () => {
    openSearchQueryStub.returns({
      error: {
        root_cause: [
          {
            type: "index_not_found_exception",
            reason: "no such index [projekti]",
            index: "projekti",
            "resource.id": "projekti",
            "resource.type": "index_or_alias",
            index_uuid: "_na_",
          },
        ],
        type: "index_not_found_exception",
        reason: "no such index [projekti]",
        index: "projekti",
        "resource.id": "projekti",
        "resource.type": "index_or_alias",
        index_uuid: "_na_",
      },
      status: 404,
    });
    await projektiSearchService.searchByOid(["1"]);
  });

  it("should handle successful result", async () => {
    openSearchQueryStub.returns({
      took: 8,
      timed_out: false,
      _shards: { total: 5, successful: 5, skipped: 0, failed: 0 },
      hits: {
        total: { value: 2, relation: "eq" },
        max_score: 1.0,
        hits: [
          {
            _index: "projekti",
            _type: "_doc",
            _id: "1",
            _score: 1.0,
            _source: {
              muistiinpano: "testi",
              tyyppi: "TIE",
              velho: {
                tilaajaOrganisaatio: "Etelä-Pohjanmaan ELY-keskus",
                nimi: "Nimi1",
                vaylamuoto: ["tie"],
              },
              status: "EI_JULKAISTU",
            },
          },
          {
            _index: "projekti",
            _type: "_doc",
            _id: "2",
            _score: 1.0,
            _source: {
              muistiinpano: "",
              tyyppi: "TIE",
              velho: {
                tilaajaOrganisaatio: "Uudenmaan ELY-keskus",
                nimi: "Nimi2",
                vaylamuoto: ["tie"],
              },
              status: "EI_JULKAISTU",
            },
          },
        ],
      },
    });
    expect(await projektiSearchService.searchByOid(["1", "2"])).toMatchSnapshot();
  });

  it("should handle query parameters successfully", async () => {
    openSearchQueryStub.onFirstCall().returns(fakeSearchResponse);
    openSearchQueryStub.onSecondCall().returns({
      aggregations: {
        aktiivinen: {
          buckets: [
            {
              key: 0,
              doc_count: 0,
              projektiTyyppi: {
                buckets: [],
              },
            },
            {
              key: 1,
              doc_count: 1,
              projektiTyyppi: {
                buckets: [
                  {
                    key: ProjektiTyyppi.TIE,
                    doc_count: 1,
                  },
                ],
              },
            },
          ],
        },
      },
    });

    await projektiSearchService.searchYllapito({
      projektiTyyppi: ProjektiTyyppi.TIE,
      nimi: "foo",
      vaylamuoto: ["tie"],
      maakunta: kuntametadata.idsForMaakuntaNames(["Pirkanmaa"]),
      suunnittelustaVastaavaViranomainen: [Viranomainen.VAYLAVIRASTO, Viranomainen.UUDENMAAN_ELY],
      vaihe: [Status.EI_JULKAISTU, Status.SUUNNITTELU],
    });
    expect(openSearchQueryStub.getCalls()[0].args[0]).toMatchSnapshot();
    expect(openSearchQueryStub.getCalls()[1].args[0]).toMatchSnapshot();
  });

  it("should handle query parameters successfully for public search", async () => {
    openSearchSuomiQueryStub.returns(fakeSearchResponse);

    await projektiSearchService.searchJulkinen({
      kieli: Kieli.SUOMI,
    });
    expect(openSearchSuomiQueryStub.getCalls()[0].args[0]).toMatchSnapshot();
  });
});
