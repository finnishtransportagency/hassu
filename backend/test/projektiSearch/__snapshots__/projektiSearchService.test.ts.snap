// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjektiSearchService should handle query parameters successfully 1`] = `
Object {
  "from": 0,
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "bool": Object {
            "must": Array [
              Object {
                "prefix": Object {
                  "nimi": "foo",
                },
              },
              Object {
                "prefix": Object {
                  "nimi": "bar",
                },
              },
            ],
          },
        },
        Object {
          "terms": Object {
            "maakunnat": Array [
              6,
            ],
          },
        },
        Object {
          "terms": Object {
            "vaylamuoto.keyword": Array [
              "tie",
            ],
          },
        },
        Object {
          "terms": Object {
            "vaihe.keyword": Array [
              "EI_JULKAISTU",
              "SUUNNITTELU",
              "EI_JULKAISTU_PROJEKTIN_HENKILOT",
            ],
          },
        },
        Object {
          "terms": Object {
            "suunnittelustaVastaavaViranomainen.keyword": Array [
              "VAYLAVIRASTO",
              "UUDENMAAN_ELY",
            ],
          },
        },
        Object {
          "term": Object {
            "projektiTyyppi.keyword": "TIE",
          },
        },
        Object {
          "term": Object {
            "aktiivinen": true,
          },
        },
      ],
    },
  },
  "size": 10,
  "sort": Array [
    Object {
      "paivitetty": Object {
        "order": "desc",
      },
    },
    Object {
      "nimi.keyword": Object {
        "order": "asc",
      },
    },
  ],
}
`;

exports[`ProjektiSearchService should handle query parameters successfully 2`] = `
Object {
  "aggs": Object {
    "aktiivinen": Object {
      "aggs": Object {
        "projektiTyyppi": Object {
          "terms": Object {
            "field": "projektiTyyppi.keyword",
            "size": 10,
          },
        },
      },
      "terms": Object {
        "field": "aktiivinen",
        "size": 10,
      },
    },
  },
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "bool": Object {
            "must": Array [
              Object {
                "prefix": Object {
                  "nimi": "foo",
                },
              },
              Object {
                "prefix": Object {
                  "nimi": "bar",
                },
              },
            ],
          },
        },
        Object {
          "terms": Object {
            "maakunnat": Array [
              6,
            ],
          },
        },
        Object {
          "terms": Object {
            "vaylamuoto.keyword": Array [
              "tie",
            ],
          },
        },
        Object {
          "terms": Object {
            "vaihe.keyword": Array [
              "EI_JULKAISTU",
              "SUUNNITTELU",
              "EI_JULKAISTU_PROJEKTIN_HENKILOT",
            ],
          },
        },
        Object {
          "terms": Object {
            "suunnittelustaVastaavaViranomainen.keyword": Array [
              "VAYLAVIRASTO",
              "UUDENMAAN_ELY",
            ],
          },
        },
      ],
    },
  },
}
`;

exports[`ProjektiSearchService should handle query parameters successfully for public search 1`] = `
Object {
  "from": 0,
  "query": Object {
    "bool": Object {
      "must": Array [
        Object {
          "range": Object {
            "publishTimestamp": Object {
              "lte": "now",
            },
          },
        },
      ],
    },
  },
  "size": 10,
  "sort": Array [
    Object {
      "viimeisinJulkaisu": Object {
        "order": "desc",
      },
    },
    Object {
      "nimi.keyword": Object {
        "order": "asc",
      },
    },
  ],
}
`;

exports[`ProjektiSearchService should handle successful result 1`] = `
Array [
  Object {
    "aktiivinen": true,
    "asiatunnus": "TESTI.123",
    "muokkaajat": Array [
      "A000112",
      "LX1",
    ],
    "nimi": "Maantie 370 Ravikylän jkpp-tie ja liittymien parantaminen",
    "oid": "1",
    "paivitetty": "2023-02-15T09:48:37+02:00",
    "projektiTyyppi": "TIE",
    "projektipaallikko": "Pertti Paallikko",
    "saame": false,
    "suunnittelustaVastaavaViranomainen": "KAAKKOIS_SUOMEN_ELY",
    "vaihe": "EI_JULKAISTU",
    "vaylamuoto": Array [
      "tie",
    ],
  },
  Object {
    "aktiivinen": true,
    "asiatunnus": "PIRELY/8634/2019",
    "maakunnat": Array [
      6,
    ],
    "muokkaajat": Array [
      "A000112",
    ],
    "nimi": "Maantie 249 parantaminen välillä Kaukolantien liittymä – Pehtoorinpolun liittymä, tie- ja rakennussuunnitelma, Sastamala",
    "oid": "2",
    "paivitetty": "2023-04-19T16:40:36+03:00",
    "projektiTyyppi": "TIE",
    "projektipaallikko": "Pekka Paallikko",
    "saame": false,
    "suunnittelustaVastaavaViranomainen": "PIRKANMAAN_ELY",
    "vaihe": "SUUNNITTELU",
    "vaylamuoto": Array [
      "tie",
    ],
    "viimeisinJulkaisu": "2023-02-17",
  },
]
`;
